pageserver/src/page_service.rs

```diff
@@ -576,6 +576,29 @@ impl Connection {
                 self.stream.write_u32::<BE>(4 + data.len() as u32)?;
                 self.stream.write_all(&data)?;
             }
+
+            BeMessage::ErrorResponse(error_msg) => {
+                let severity = Bytes::from("SERROR\0");
+                let code = Bytes::from("CXX000\0");
+
+                self.stream.write_u8(b'E')?;
+                self.stream.write_u32::<BE>(
+                    4 + severity.len() as u32
+                        + code.len() as u32
+                        + (1 + error_msg.len() as u32 + 1)
+                        + 1,
+                )?;
+
+                self.stream.write_all(&severity)?;
+
+                self.stream.write_all(&code)?;
+
+                self.stream.write_u8(b'M')?; /* primary human-readable error message */
+                self.stream.write_all(error_msg.as_bytes())?;
+                self.stream.write_u8(0)?; /* end of M field */
+
+                self.stream.write_u8(0)?; /* end of all fields */
+            }
         }
```

Comment:
Can we add a comment before each network message, describing the message being constructed and why?

It's helpful to those of us who don't know the postgres network protocol very well.

Reply:
Message names here correspond to names in this doc:
https://www.postgresql.org/docs/13/protocol-message-formats.html

We can put that somewhere in the header.

Reply:
I would still prefer a comment at the start of each message.
- It tells me these are two separate messages, and not one message nested inside another
- Why are we sending this message?

Sorry if this is really obvious to people who know postgres better than I do. To me this code seems not very clear. This example probably isn't that bad, but there are other places where it's much harder to understand the meaning.

In general, I think we can do much better than manually encoding/decoding things; I would like to see a struct for each message, and low-level encoding tasks like null-terminating strings should be handled by a `PqString` type that knows how to encode/decode to/from a Rust string.

Reply:
+1 to better comments, and for refactoring this encoding/decoding stuff and constructing the message to separate structs. And probably move those to a separate source file. And share the code with walkeeper. And also a more principle way of parsing the commands we have, with a real grammar instead of regexps.

Those thing don't need to block this PR, though. This is just following the precedent, albeit bad one.

Reply:
I described message format in comments.

Other way of parsing can be done within https://github.com/zenithdb/zenith/issues/118 -- IMO that can be quite useful public crate, as we are not the only database that pretends to be postgres.